source/s3_client.c

```diff
@@ -1482,39 +1486,234 @@ static void s_s3_client_on_acquire_http_connection(
 }
 
-/* Called by aws_s3_meta_request when it has finished using this VIP connection for a single request. */
-void aws_s3_client_notify_connection_finished(
-    struct aws_s3_client *client,
-    struct aws_s3_vip_connection *vip_connection) {
+static void s_s3_client_retry_with_token(
+    struct aws_retry_strategy *retry_strategy,
+    int error_code,
+    struct aws_retry_token *token,
+    void *user_data) {
+
+    AWS_PRECONDITION(retry_strategy);
+    (void)retry_strategy;
+
+    struct aws_s3_vip_connection *vip_connection = user_data;
+    AWS_PRECONDITION(vip_connection);
+    AWS_PRECONDITION(vip_connection->owning_vip);
+
+    struct aws_s3_client *client = vip_connection->owning_vip->owning_client;
     AWS_PRECONDITION(client);
+
+    struct aws_s3_request *request = vip_connection->request;
+    AWS_PRECONDITION(request);
+
+    struct aws_s3_meta_request *meta_request = request->meta_request;
+    AWS_PRECONDITION(meta_request);
+
+    if (error_code != AWS_ERROR_SUCCESS) {
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_META_REQUEST,
+            "id=%p Meta request could not get retry token for request %p due to error %d (%s)",
+            (void *)meta_request,
+            (void *)request,
+            error_code,
+            aws_error_str(error_code));
+
+        goto error_clean_up;
+    }
+
+    AWS_ASSERT(token);
+
+    enum aws_retry_error_type error_type = AWS_RETRY_ERROR_TYPE_TRANSIENT;
+
+    switch (request->send_data.error_code) {
+        case AWS_ERROR_S3_INTERNAL_ERROR:
+            error_type = AWS_RETRY_ERROR_TYPE_SERVER_ERROR;
+            break;
+
+        case AWS_ERROR_S3_SLOW_DOWN:
+            error_type = AWS_RETRY_ERROR_TYPE_THROTTLING;
+            break;
+    }
+
+    vip_connection->retry_token = token;
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_S3_CLIENT,
+        "id=%p Client scheduling retry of request %p for meta request %p with token %p.",
+        (void *)client,
+        (void *)request,
+        (void *)meta_request,
+        (void *)token);
+
+    /* Ask the retry strategy to schedule a retry of the request. */
+    if (aws_retry_strategy_schedule_retry(
+            vip_connection->retry_token, error_type, s_s3_client_retry_ready, vip_connection)) {
+        error_code = aws_last_error_or_unknown();
+
+        AWS_LOGF_ERROR(
+            AWS_LS_S3_CLIENT,
+            "id=%p Client could not retry request %p for meta request %p with token %p due to error %d (%s)",
+            (void *)client,
+            (void *)request,
+            (void *)meta_request,
+            (void *)token,
+            error_code,
+            aws_error_str(error_code));
+
+        goto error_clean_up;
+    }
+
+    return;
+
+error_clean_up:
+
+    request->send_data.error_code = error_code;
+    aws_s3_client_notify_connection_finished(client, vip_connection, 0);
+}
+
+static void s_s3_client_retry_ready(struct aws_retry_token *token, int error_code, void *user_data) {
```

Comment:
Isn't this more `retry_complete` than `retry_ready`?

Reply:
no it's retry ready, it means, "retry now please"